File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/ThreadLocalRandomWrapperMersenneTwisterTest.java
package io.github.pr0methean.betterrandom.prng;

import io.github.pr0methean.betterrandom.seed.SeedException;
import io.github.pr0methean.betterrandom.seed.SeedGenerator;
import io.github.pr0methean.betterrandom.util.SerializableSupplier;
import java.util.Map;
import java8.util.function.Function;
import java8.util.function.Supplier;
import org.testng.annotations.Test;

@Test(testName = "ThreadLocalRandomWrapper:MersenneTwisterRandom")
public class ThreadLocalRandomWrapperMersenneTwisterTest extends ThreadLocalRandomWrapperTest {

  private final MersenneTwisterRandomColonColonNew mtSupplier
      = new MersenneTwisterRandomColonColonNew(getTestSeedGenerator());

  @Override public Map<Class<?>, Object> constructorParams() {
    final Map<Class<?>, Object> params = super.constructorParams();
    params.put(Supplier.class, mtSupplier);
    params.put(Function.class, mtSupplier);
    return params;
  }

  @Override protected BaseRandom createRng() throws SeedException {
    return new ThreadLocalRandomWrapper(mtSupplier);
  }

  protected static class MersenneTwisterRandomColonColonNew
      implements SerializableSupplier<MersenneTwisterRandom>, Function<byte[], BaseRandom> {

    private final SeedGenerator seedGenerator;

    public MersenneTwisterRandomColonColonNew(SeedGenerator seedGenerator) {
      this.seedGenerator = seedGenerator;
    }

    @Override public MersenneTwisterRandom get() {
      return new MersenneTwisterRandom(seedGenerator);
    }

    @Override public BaseRandom apply(byte[] seed) {
      return new MersenneTwisterRandom(seed);
    }
  }
}

File: betterrandom/src/test/java/io/github/pr0methean/betterrandom/prng/ThreadLocalRandomWrapperMersenneTwisterTest.kt
package io.github.pr0methean.betterrandom.prng

import io.github.pr0methean.betterrandom.seed.SeedException
import io.github.pr0methean.betterrandom.seed.SeedGenerator
import java.io.Serializable
import java.util.function.Function
import java.util.function.Supplier
import org.testng.annotations.Test

@Test(testName = "ThreadLocalRandomWrapper:MersenneTwisterRandom")
open class ThreadLocalRandomWrapperMersenneTwisterTest : ThreadLocalRandomWrapperTest() {

    private val mtSupplier: Supplier<out BaseRandom>

    init {
        // Must be done first, or else lambda won't be serializable.
        val seedGenerator = testSeedGenerator

        mtSupplier = { MersenneTwisterRandom(seedGenerator) } as Serializable
    }

    override fun constructorParams(): Map<Class<*>, Any> {
        val params = super.constructorParams()
        params[Supplier<*>::class.java] = mtSupplier
        params[Function<*, *>::class.java] = Function<ByteArray, BaseRandom> { MersenneTwisterRandom(it) }
        return params
    }

    @Throws(SeedException::class)
    override fun createRng(): BaseRandom {
        return ThreadLocalRandomWrapper(mtSupplier)
    }
}
